### PR TITLE
Feature/script for uploading dialog flow

### PR DESCRIPTION
**Pull request by bimlendra**
#33 
Script for uploading dialogflow

### CODE REVIEW

File: functions/uploader/entities/entities.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+const error = require(`debug`)(`ia:uploader:entities:error`);
+const util = require(`util`);
+const mustache = require('mustache');
+const config = require('../config');
+const MAX_ENTITY = 30000;
+const MAX_REQUEST = 1000;
+
+const basicHeaderRequest = {
+  'content-type': 'application/json; charset=UTF-8',
+  'authorization': 'BEARER a55db7f7ed0a4b52a0982b22e2428f18'
```
[the remark this answers is not in the document]
Comment (bimlendra) [reply functions/uploader/entities/entities.js]: @eugene: Can you suggest me a place where you have used process.env? Actually, if I set it using code then I will have to include the value in code. So, I want to see how you are implementing this.

**Comment by bimlendra** [comment]
@hyzhak : Not sure, why these tests are getting fail. Can you help?

**Comment by hyzhak** [comment]
@bimlendra you're fetching one URL, but mocking another. Just compare carefully

**Comment by bimlendra** [comment]
Just modified that. Actually, it was changed in master. Since, I took last pull. Pushed again after making changes.

File: functions/uploader/entities/entities.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hyzhak) [reply functions/uploader/entities/entities.js]: I don't have such places yet. And you can read it from any node.js file. But you shouldn't set it.
you can set it in your console:
```
export DEBUG=*
```
and use everywhere, or
```
DEBUG=* npm test
```

[the remark this answers is not in the document]
Comment (bimlendra) [reply functions/uploader/entities/entities.js]: But in that case, I guess everyone will have to set  that in their console.

File: functions/tests/uploader/entities/entities.spec.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+        });
+
+        it('should fetch entity information from DF', () => {
+          return entities.fetchEntitiesFromDF('testing-collection')
+            .then(items => {
+              for (let i = 0; i < items.length; i++) {
+                expect(items[i]).to.be.a('string');
+              }
+            })
+            .catch((err) => {
```
Comment (hyzhak) [inline functions/tests/uploader/entities/entities.spec.js]: the same as above - choose one and test

File: functions/tests/uploader/entities/ia-request.spec.js [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+              fetchMock
+                .sandbox()
+                .get('begin:https://web.archive.org/advancedsearch.php?q=collection:(etree)', getCollectionFromIA)
+            );
+          });
+
+          it('should fetch collection from IA', () => {
+            iaRequest.fetchEntitiesFromIA(`etree`, `10`)
+              .then(items => {
+                console.log(util.inspect(items, false, null));
```
Comment (hyzhak) [inline functions/tests/uploader/entities/ia-request.spec.js]: do you really need console.log here?

File: functions/tests/uploader/index.spec.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        expect(index.handle).to.be.ok;
+      });
+    });
+    describe('execute', () => {
+      it('should be defined', () => {
+        expect(index.execute).to.be.ok;
+      });
+      it('should reject for unknown command', () => {
+        const condition = index.execute(`anything`, `anything`);
+        //  postEntitiesToDF(entityname, dif, 0);
```
Comment (hyzhak) [inline functions/tests/uploader/index.spec.js]: do you need this comment?

File: functions/uploader/entities/entities.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (hyzhak) [reply functions/uploader/entities/entities.js]: Sure, if someone would like to work on behalf of the app they should get private key or permission to do so. We don't have the goal to allow everyone in Web to upload any entities which they want to our Dialog Flow app.

File: functions/uploader/entities/entities.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+      error(`Get error in fetching entity from DF, error: ${JSON.stringify(e)}`);
+      return Promise.reject(e);
+    });
+}
+function deleteAllEntitiesFromDF (entityname) {
+  return Promise.all([
+    fetchEntitiesFromDF(entityname),
+  ])
+    .then(entities => {
+      var spliceCreators = entities.splice(0, 1);
```
Comment (hyzhak) [inline functions/uploader/entities/entities.js]: do you really need to splice first single element only?

File: functions/uploader/entities/entities.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (bimlendra) [reply functions/uploader/entities/entities.js]: Got your point now.

[the remark this answers is not in the document]
Comment (hyzhak) [reply functions/uploader/entities/entities.js]: @bimlendra btw, please find documentation - how could we revoke those keys

File: functions/config.js [hunk 6]
```diff
@@ -24,6 +24,9 @@ module.exports = {
     '&page={{page}}' +
     '&output=json',
     SONG_URL: 'https://archive.org/download/{{albumId}}/{{filename}}',
+    DF_ENTITY_GET_URL: 'https://api.dialogflow.com/v1/entities/{{entityname}}?v=20150910',
```
[the remark this answers is not in the document]
Comment (bimlendra) [reply functions/config.js]: Off-course not, will remove that.

File: functions/uploader/entities/entities.js [hunk 7]
```diff
@@ -10,12 +10,20 @@ const MAX_ENTITY = 30000;
 const MAX_REQUEST = 1000;
 
 const basicHeaderRequest = {
-  'content-type': 'application/json; charset=UTF-8',
-  'authorization': 'BEARER a55db7f7ed0a4b52a0982b22e2428f18'
+  `content-type`: `application/json; charset=UTF-8`,
+  `authorization`: `BEARER ${process.env.AUTHORIZATION}`
```
Comment (hyzhak) [inline functions/uploader/entities/entities.js]: `AUTHORIZATION` is too general we should use something more specific, for example: `DIALOG_FLOW_DEV_TOKEN`

File: functions/uploader/entities/entities.js [hunk 8]
```diff
@@ -66,6 +74,12 @@ function postEntitiesToDF (entityname, entities, first) {
     });
 }
 
+/**
+ * Get entities from DialogFlow
```
Comment (hyzhak) [inline functions/uploader/entities/entities.js]: Do we get entities or entity here? `@param entityname` but `fetchEntities...`

**Comment by bimlendra** [comment]
Not required anymore.

**Pull request closed without merging**